pyserini/search/__main__.py

```diff
@@ -130,37 +137,68 @@
+    if use_prcl and len(hits) > (args.r + args.n):
+        scores, docids = ranker.rerank(docids, scores)
 
-            if args.msmarco:
-                for i, docid in enumerate(docids):
-                    target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
-            else:
-                for i, (docid, score) in enumerate(zip(docids, scores)):
-                    target_file.write(f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+    if args.msmarco:
+        for i, docid in enumerate(docids):
+            target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
+    else:
+        for i, (docid, score) in enumerate(zip(docids, scores)):
+            target_file.write(
+                f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+
+
+def write_result_max_passage(result: Tuple[str, List[JSimpleSearcher]]):
+    topic, hits = result
+    unique_docs = set()
+    rank = 1
+    for hit in hits:
+        docid, _ = hit.docid.split(args.max_passage_delimiter)
+        if docid in unique_docs:
+            continue
+
+        if args.msmarco:
+            target_file.write(f'{topic}\t{docid}\t{rank}\n')
+        else:
+            target_file.write(
+                f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
+        rank = rank + 1
+        unique_docs.add(docid)
+        if rank > args.max_passage_hits:
+            break
+
+
+with open(output_path, 'w') as target_file:
+    batch_topics = list()
+    for index, topic in enumerate(tqdm(sorted(topics.keys()))):
+        if args.batch_size <= 1 and args.threads <= 1:
+            search = topics[topic].get('title')
+            hits = searcher.search(search, args.hits)
+            results = [(topic, hits)]
         else:
-            unique_docs = set()
-            rank = 1
-            for hit in hits:
-                docid, _ = hit.docid.split(args.max_passage_delimiter)
-                if docid in unique_docs:
-                    continue
-
-                if args.msmarco:
-                    target_file.write(f'{topic}\t{docid}\t{rank}\n')
-                else:
-                    target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
-                rank = rank + 1
-                unique_docs.add(docid)
-                if rank > args.max_passage_hits:
-                    break
+            batch_topics.append(topic)
+            if (index + 1) % args.batch_size == 0 or index == len(topics.keys()) - 1:
+                queries = [topics[topic].get('title')
+                           for topic in batch_topics]
+                ids = [str(topic) for topic in batch_topics]
+                results = searcher.batch_search(
+                    queries, ids, args.hits, args.threads)
+                results = sorted(
+                    results.items(), key=lambda item: int(item[0]))
+                batch_topics.clear()
+            else:
+                continue
+
+        for result in results:
+            if not args.max_passage:
```

Comment:
I think the if clause should be flipped... i.e., `if args.max_passage`... `else`